From 99bf2251c17e39d22f4cb5109f099eb9b36cbdd1 Mon Sep 17 00:00:00 2001
From: Jens Axboe <axboe@fb.com>
Date: Mon, 25 Apr 2016 10:36:19 -0600
Subject: blk-wb: check sync issue latency before valid stat sample

If the device is overwhelmed with writes, then we can get into
a situation where we have no valid samples. Hence we return
UNKNOWN, and we potentially ramp up the writes.

Check the sync issue latency first, so we can return EXCEEDED
if that is the case.

Signed-off-by: Jens Axboe <axboe@fb.com>
---
 block/blk-wb.c | 24 ++++++++++++------------
 1 file changed, 12 insertions(+), 12 deletions(-)

diff --git a/block/blk-wb.c b/block/blk-wb.c
index a88cc56..8410c94 100644
--- a/block/blk-wb.c
+++ b/block/blk-wb.c
@@ -191,18 +191,6 @@ static int __latency_exceeded(struct rq_wb *rwb, struct blk_rq_stat *stat)
 {
 	u64 thislat;
 
-	if (!stat_sample_valid(stat))
-		return LAT_UNKNOWN;
-
-	/*
-	 * If the 'min' latency exceeds our target, step down.
-	 */
-	if (stat[0].min > rwb->min_lat_nsec) {
-		trace_block_wb_lat(stat[0].min);
-		trace_block_wb_stat(stat);
-		return LAT_EXCEEDED;
-	}
-
 	/*
 	 * If our stored sync issue exceeds the window size, or it
 	 * exceeds our min target AND we haven't logged any entries,
@@ -215,6 +203,18 @@ static int __latency_exceeded(struct rq_wb *rwb, struct blk_rq_stat *stat)
 		return LAT_EXCEEDED;
 	}
 
+	if (!stat_sample_valid(stat))
+		return LAT_UNKNOWN;
+
+	/*
+	 * If the 'min' latency exceeds our target, step down.
+	 */
+	if (stat[0].min > rwb->min_lat_nsec) {
+		trace_block_wb_lat(stat[0].min);
+		trace_block_wb_stat(stat);
+		return LAT_EXCEEDED;
+	}
+
 	if (rwb->scale_step)
 		trace_block_wb_stat(stat);
 
-- 
cgit v0.11.2

